docs/dapps/build-dapp-from-scratch.md

```diff
@@ -0,0 +1,548 @@
+    // Ralph does not allow underflow.
+    balance = balance - amount
+
+    // Emit the event defined earlier.
+    emit Withdraw(callerAddress!(), amount)
+}
+```
+
+With the `assert!` we make sure no one takes more than 2 tokens at the same time.  
+`transferTokenFromSelf` will actually perform the transfer of the tokens.  
+We update the `mut balance` field with the new balance. In the case of underflow, an error will be raised and the transaction won't be performed.
+`callerAddress!()` and `selfTokenId!()` are built-in functions, you can read more about them in our [built-in functions page](/ralph/built-in-functions).
+## Compile your contract
+
+The compiler needs to contact the full node in order to compile the contract, you'll need to use the right information defined while [creating your devnet](/full-node/devnet). If you haven't start it, now it's the time.
+We define the node URL using the following config file: `alephium.config.ts`. 
+Create this file in the root directory of your project and paste the following code:
+
+```typescript
+import { Configuration } from '@alephium/cli'
+
+export type Settings = {}
+
+const configuration: Configuration<Settings> = {
+  defaultNetwork: 'devnet',
+  networks: {
+    devnet: {
+      //Make sure the two values match what's in your devnet configuration
+      nodeUrl: 'http://localhost:22973',
+      networkId: 2
+    }
+  }
+}
+
+export default configuration
+```
+
+Now, let's compile:
+
+```sh
+npx @alephium/cli@latest compile
+```
+
+It may ask you for some confirmation to install the latest `@alephium/cli` package. Select yes to proceed.
+
+Once the above command succeeds, you will notice that a new folder called `artifacts` was created. It contains several files related to your contract. For example, `artifacts/ts/TokenFaucet.ts` produces lots of helper functions like `at`, `fetchState`, `call*`, etc, as well as many test functions.
+
+## Test your contract
+The SDK provides unit test functionalities, which call the contract by sending a transaction, but instead of changing the blockchain state, it returns the new contract state, transaction outputs, and events.
+
+Install the test framework:
+
+```sh
+npm install ts-jest @types/jest
+```
+
+You'll also need our `@alephium/web3` package:
+
+```sh
+npm install @alephium/web3 @alephium/web3-test
+```
+
+Create a `test` folder:
+
+```sh
+mkdir test
+```
+
+and create the `test/token.test.ts` minimalistic test file with the following contents:
+
+```typescript
+import { web3, Project, addressFromContractId } from '@alephium/web3'
+import { randomContractId, testAddress } from '@alephium/web3-test'
+import { TokenFaucet } from '../artifacts/ts'
+
+describe('unit tests', () => {
+  it('Withdraws 1 token from TokenFaucet', async () => {
+
+    // Use the correct host and port
+    web3.setCurrentNodeProvider('http://127.0.0.1:22973')
+    await Project.build()
+
+    let testContractId = randomContractId()
+    let testParams = {
```

Comment:
It seems that these two variables can be const.

Reply:
Indeed, I copy/pasted from https://github.com/alephium/nextjs-template/blob/f1deb248580508052e1cd4791c604487bfb28e4a/test/token.test.ts#L7-L10 but in my simpler version I don't override :+1: